cpp/oneapi/dal/backend/interop/table_conversion.hpp

```diff
@@ -75,51 +95,63 @@ inline daal::data_management::NumericTablePtr wrap_by_host_homogen_adapter(
-    auto rows = row_accessor<const Float>{ table }.pull();
-    return convert_to_daal_homogen_table(rows, table.get_row_count(), table.get_column_count());
-}
-
-template <typename Float>
+template <typename Data>
 inline daal::data_management::NumericTablePtr convert_to_daal_table(const homogen_table& table) {
-    auto wrapper = wrap_by_host_homogen_adapter(table);
-    if (!wrapper) {
-        return copy_to_daal_homogen_table<Float>(table);
-    }
-    else {
+    if (auto wrapper = wrap_by_host_homogen_adapter(table)) {
         return wrapper;
     }
+    return copy_to_daal_homogen_table<Data>(table);
 }
 
-template <typename Float>
+template <typename Data>
 inline daal::data_management::NumericTablePtr convert_to_daal_table(const table& table) {
     if (table.get_kind() == homogen_table::kind()) {
         const auto& homogen = static_cast<const homogen_table&>(table);
-        return convert_to_daal_table<Float>(homogen);
+        return convert_to_daal_table<Data>(homogen);
     }
     else {
-        return copy_to_daal_homogen_table<Float>(table);
+        return copy_to_daal_homogen_table<Data>(table);
     }
 }
 
 #ifdef ONEDAL_DATA_PARALLEL
-template <typename T>
-inline auto convert_to_daal_sycl_homogen_table(sycl::queue& queue,
-                                               array<T>& data,
-                                               std::int64_t row_count,
-                                               std::int64_t column_count) {
-    data.need_mutable_data(queue);
-    const auto daal_data =
-        daal::services::SharedPtr<T>(data.get_mutable_data(), daal_object_owner{ data });
+inline daal::data_management::NumericTablePtr convert_to_daal_table(const sycl::queue& queue,
+                                                                    const table& table) {
+    return interop::sycl_table_adapter::create(queue, table);
+}
 
+template <typename Data>
+inline daal::data_management::NumericTablePtr convert_to_daal_table(const sycl::queue& queue,
+                                                                    const array<Data>& data,
+                                                                    std::int64_t row_count,
+                                                                    std::int64_t column_count) {
+    using daal::services::Status;
+    using daal::services::SharedPtr;
+    using daal::services::internal::Buffer;
     using daal::data_management::internal::SyclHomogenNumericTable;
-    return SyclHomogenNumericTable<T>::create(daal_data,
-                                              dal::detail::integral_cast<std::size_t>(column_count),
-                                              dal::detail::integral_cast<std::size_t>(row_count),
-                                              queue);
-}
+    using dal::detail::integral_cast;
 
+    ONEDAL_ASSERT(data.get_count() == row_count * column_count);
+    ONEDAL_ASSERT(data.has_mutable_data());
```

Comment:
There should be an assert that array contains USM data and `queue` is compatible with `data.get_queue()` I guess